ruby_event_store/lib/ruby_event_store/mappers/protobuf.rb

```diff
@@ -0,0 +1,28 @@
+module RubyEventStore
+  module Mappers
+    class Protobuf
+      def initialize(event_id_getter: :event_id, events_class_remapping: {})
+        @event_id_getter = event_id_getter
+        @events_class_remapping = events_class_remapping
+      end
+
+      def event_to_serialized_record(domain_event)
+        SerializedRecord.new(
+          event_id:   domain_event.public_send(event_id_getter),
```

Comment:
If there is an assumption that the Protobuf message would self-contain all the event data, along with the event_id (possibly) and metadata (possibly) is there a real value in trying to extract the ID from the message? The moment you agree to accept an event which structure is beyond your control you should not make any assumptions about this structure. 
What if the message would not really have any kind of event_id? One may try to pass one of the message attributes as an ID just to fulfil the mapper's requirement. And what if the attribute is not unique?
Moreover, the event_id is skipped when deserializing the record back to the protobuf message (it's obvious - the message itself is the event in this case).

Having said that, I'd choose ensuring the id is unique over trying extracting the id from the message data:
```ruby
event_id:  SecureRandom.uuid
```

Reply:
Hm, the moment I submitted the comment I realised it still must be possible to retrieve the event from the stream by ID. So, maybe some kind of a checksum would serve as a better event id than a random string?

Reply:
@siemakuba If the structure if your messages does not fit the assumptions of our mapper and repository standard then you can substitute the mapper and repository to something that you implement on your own.

> is there a real value in trying to extract the ID from the message?

yes, we can have a reference to it outside the message, in a DB or in-memory.

> What if the message would not really have any kind of event_id

Then you cannot persist this message with our default repository anyway, so this mapper is useless to you anyway :)

> I'd choose ensuring the id is unique over trying extracting

https://github.com/RailsEventStore/rails_event_store/blob/f866e07c046f2b67443b89913edc485318a538eb/rails_event_store_active_record/lib/rails_event_store_active_record/generators/templates/migration_template.rb#L19

Doing that for our repository requires having access to it.

I was thinking about using random uuid which would not be inside the message. But then whoever got that message would not know the ID because it would not be there so it would not know what message it is processing and would not be capable of doing exactly once processing so I realized _meh, the messages must contain the event_id anyway_. It's harder with _metadata_ but that's a bigger take.

Imho we should in the future switch types of messages for writing and for reading. Read messages probably can contain more data than just what's been sent. It could contain its position in a stream, stream_name, whether that's the original stream it was written first or not etc. But that's not where we are right now. So in the future I see this as some kind of Container+Original message.

> maybe some kind of a checksum would serve as a better event id than a random string

how can you compute a checksum of a message which must contain the value itself?